File: src/main/java/xyz/nkomarn/composter/network/protocol/packet/s2c/ClientboundTeleportEntityPacket.java
package xyz.nkomarn.composter.network.protocol.packet.s2c;

import io.netty.buffer.ByteBuf;
import io.netty.buffer.Unpooled;
import org.jetbrains.annotations.NotNull;
import xyz.nkomarn.composter.entity.Entity;
import xyz.nkomarn.composter.network.protocol.Packet;

import static xyz.nkomarn.composter.util.ByteBufUtil.toAbsoluteInteger;

public class ClientboundTeleportEntityPacket extends Packet<ClientboundTeleportEntityPacket> {

    private int id;
    private int x;
    private int y;
    private int z;
    private byte yaw;
    private byte pitch;

    public ClientboundTeleportEntityPacket() {
    }

    public ClientboundTeleportEntityPacket(Entity entity) {
        this.id = entity.getId();

        var pos = entity.getPos();
        this.x = toAbsoluteInteger(pos.getX());
        this.y = toAbsoluteInteger(pos.getY());
        this.z = toAbsoluteInteger(pos.getZ());
        this.yaw = (byte) (entity.getYaw() * 256F / 360F);
        this.pitch = (byte) (entity.getPitch() * 256F / 360F);

        System.out.printf("(%s, %s, %s)\n", x, y, z);
    }

    @Override
    public int getId() {
        return 0x22;
    }

    @Override
    public @NotNull ByteBuf encode() {
        return Unpooled.buffer()
                .writeInt(id)
                .writeInt(x)
                .writeInt(y)
                .writeInt(z)
                .writeByte(yaw)
                .writeByte(pitch);
    }
}

File: src/main/kotlin/kyta/composter/protocol/packet/play/ClientboundTeleportEntityPacket.kt
package kyta.composter.protocol.packet.play

import kyta.composter.protocol.asRotation
import kyta.composter.math.Vec3d
import kyta.composter.protocol.Packet
import kyta.composter.protocol.PacketSerializer
import kyta.composter.protocol.WriteBuffer

class ClientboundTeleportEntityPacket(
    val id: Int,
    val pos: Vec3d,
    val yaw: Float,
    val pitch: Float,
) : Packet {
    companion object : PacketSerializer<ClientboundTeleportEntityPacket> {
        override fun serialize(packet: ClientboundTeleportEntityPacket, buffer: WriteBuffer) {
            buffer.writeInt(packet.id)
            buffer.writeAbsoluteInt(packet.pos.x)
            buffer.writeAbsoluteInt(packet.pos.y)
            buffer.writeAbsoluteInt(packet.pos.z)
            buffer.writeByte(packet.yaw.asRotation())
            buffer.writeByte(packet.pitch.asRotation())
        }
    }
}